File: news_server/modules/source/infrastructure/getters/linuxdo.ts
/**
 * LINUX DO 新闻源获取器
 * 从 newsnow/server/sources/linuxdo.ts 迁移
 */
import type { SourceGetter } from '../../domain/getter/getter'
import { NewsItem } from '../../domain/newsitem/entity'

interface LinuxdoTopic {
  id: number
  title: string
  fancy_title: string
  posts_count: number
  reply_count: number
  created_at: string
  last_posted_at: string
  visible: boolean
  archived: boolean
  pinned: boolean
}

interface LinuxdoResponse {
  topic_list: {
    topics: LinuxdoTopic[]
  }
}

export const linuxdoHotGetter: SourceGetter = async () => {
  const response = await fetch('https://linux.do/top/daily.json')
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Linuxdo: ${response.statusText}`)
  }

  const res = (await response.json()) as LinuxdoResponse

  return res.topic_list.topics
    .filter((k) => k.visible && !k.archived && !k.pinned)
    .map((k) =>
      NewsItem.fromEntity({
        id: k.id.toString(),
        title: k.title,
        url: `https://linux.do/t/topic/${k.id}`,
      })
    )
}

export const linuxdoLatestGetter: SourceGetter = async () => {
  const response = await fetch('https://linux.do/latest.json?order=created')
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Linuxdo: ${response.statusText}`)
  }

  const res = (await response.json()) as LinuxdoResponse

  return res.topic_list.topics
    .filter((k) => k.visible && !k.archived && !k.pinned)
    .map((k) =>
      NewsItem.fromEntity({
        id: k.id.toString(),
        title: k.title,
        url: `https://linux.do/t/topic/${k.id}`,
        pubDate: new Date(k.created_at).getTime(),
      })
    )
}

File: modules/source/infrastructure/getters/linuxdo.go
package getters

import (
	"context"
	"time"
)

func (r *Registry) linuxdoHot(ctx context.Context) ([]Item, error) {
	return r.linuxdo(ctx, "https://linux.do/top/daily.json", false)
}

func (r *Registry) linuxdoLatest(ctx context.Context) ([]Item, error) {
	return r.linuxdo(ctx, "https://linux.do/latest.json?order=created", true)
}

func (r *Registry) linuxdo(ctx context.Context, api string, withDate bool) ([]Item, error) {
	var res struct {
		TopicList struct {
			Topics []struct {
				ID        int    `json:"id"`
				Title     string `json:"title"`
				Visible   bool   `json:"visible"`
				Archived  bool   `json:"archived"`
				Pinned    bool   `json:"pinned"`
				CreatedAt string `json:"created_at"`
			} `json:"topics"`
		} `json:"topic_list"`
	}
	if err := r.getJSON(ctx, api, nil, &res); err != nil {
		return nil, err
	}
	var out []Item
	for _, k := range res.TopicList.Topics {
		if !k.Visible || k.Archived || k.Pinned {
			continue
		}
		it := Item{ID: itoa(k.ID), Title: k.Title, URL: "https://linux.do/t/topic/" + itoa(k.ID)}
		if withDate {
			if t, err := time.Parse(time.RFC3339, k.CreatedAt); err == nil {
				it.PubDate = t.UnixMilli()
			}
		}
		out = append(out, it)
	}
	return out, nil
}
